joey/layers_v2.py

```diff
@@ -391,3 +391,158 @@ def __init__(self, kernel_size, input_size,
+        self._dims = len(input_size)-2
+        dimensions = ['dbatch', 'dchannel']
+        result_shape = []
+        # generating  in the order depth, height, width ,
+        # hence arr[-3], arr[-2] and so on
+        for i in range(0, self._dims):
+            result_d = (input_size[(-self._dims+i)] * self._scale_factor[i])
+            result_shape.append(result_d)
+            dimensions.append('d_'+dim_dict.get(self._dims-i, self._dims-i))
+
+        result_shape = (*input_size[0:2], *result_shape)
+
+        # input data function
+        input_dimensions = [SpaceDimension("Input_"+x) for x in dimensions]
+
+        input_func = Function(name="Input_F",
+                              shape=(input_size), space_order=0,
+                              dimensions=input_dimensions, dtype=np.float64)
+
+        result_dimensions = [SpaceDimension("Result_"+x) for x in dimensions]
+
+        result_func = Function(name="Result_F", shape=result_shape,
+                               dimensions=result_dimensions, space_order=0,
+                               dtype=np.float64)
+
+        return (None, input_func, result_func, None, None,
+                None, None)
+
+    def execute(self, input_data) -> np.array:
+
+        self._I.data[:] = input_data
+        self._R.data[:] = 0
+        return super().execute()
+
+    def equations(self):
+
+        input_dimensions = self._I.dimensions
+        res_dims = [input_dimensions[0], input_dimensions[1]]
+        args = []
+        for i in range(0, self._dims):
+            new_dim = SpaceDimension(
+                name='d_upsample'+self._dim_dict.get
+                (self._dims-i, self._dims-i))
+            res_dims.append(
+                input_dimensions
+                [-self._dims + i]*self._scale_factor[i]+new_dim)
+            args.append((new_dim.name + '_M', self._scale_factor[i]-1))
+
+        eqs = [Eq(self._R[res_dims], self._I[input_dimensions])]
+
+        return (eqs, args)
+
+    def backprop_equations(self, prev_layer, next_layer):
+        layer = self
+
+        kernel_dims = layer.kernel_gradients.dimensions
+        bias_dims = layer.bias_gradients.dimensions
+        dims = layer.result_gradients.dimensions
+
+        eqs = [Inc(layer.bias_gradients[bias_dims[0]],
+                   layer.result_gradients[dims[0], dims[1], dims[2], dims[3]]),
+               Inc(layer.kernel_gradients[kernel_dims[0], kernel_dims[1],
+                                          kernel_dims[2], kernel_dims[3]],
+                   layer.result_gradients[dims[0],
+                                          kernel_dims[0], dims[2],
+                                          dims[3]] *
+                   layer.input[dims[0], kernel_dims[1],
+                               kernel_dims[2] + dims[2],
+                               kernel_dims[3] + dims[3]])]
+
+        _, _, height, width = layer.kernel.shape
+
+        if next_layer is not None:
```

Comment:
I don't like the fact that we follow different code paths depending on whether the layer is stand-alone or as part of a larger network.

Reply:
This PR is only for forward propogation.
We have diff equation for back-prop but with same if statement, I continued the original flow for back-propogation as its additional code if you have a previous layer which makes sense.
from what I understand, its not about single or part of network, its whether its the first layer or not.
We can't calculate next_layer variables if next layer is none 
Do you have something else in mind?

Reply:
Following a discussion, we agree that this will go away when we separate layer specification vs layer execution.

Reply:
Issue #17